src/toxcore/dht/server/hole_punching.rs

```diff
@@ -0,0 +1,363 @@
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+/*!
+Module for hole-punching.
+
+https://zetok.github.io/tox-spec/#hole-punching
+*/
+
+use std::net::{IpAddr, SocketAddr};
+use std::time::{Duration, Instant};
+use std::collections::HashMap;
+use futures::{future, stream, Stream};
+
+use toxcore::crypto_core::*;
+use toxcore::dht::server::*;
+use toxcore::dht::packed_node::*;
+use toxcore::io_tokio::IoFuture;
+
+/// Interval in seconds for sending NatPingRequest
+pub const NAT_PING_PUNCHING_INTERVAL: u64 = 3;
+/// Interval in seconds to reset trying count of hole-punch,
+/// if connecttion to friends is not done for over 40 seconds
+/// retry to connect to friends using NatPingRequest
+pub const RESET_PUNCH_INTERVAL: u64 = 40;
+// Maximum number of ports to punch hole
+const MAX_PORTS_TO_PUNCH: u32 = 48;
+// Maximum trying count for hole punch
+// after RESET_PUNCH_INTERVAL seconds it is reset if connection is not done.
+const MAX_NORMAL_PUNCHING_TRIES: u32 = 5;
+/// Maximum clients number per friend
+pub const MAX_CLIENTS_PER_FRIEND: u32 = 8;
+
+/// Structure for hole punch
+pub struct HolePunching {
+    /// flag for hole punching is done or not.
+    pub is_punching_done: bool,
+    /// number of punching tries
+    pub num_punch_tries: u32,
+    /// last timestamp of receiving NatPingResponse packet
+    pub last_recv_ping_time: Instant,
+    /// last timestamp of sending NatPingRequest packet
+    pub last_send_ping_time: Instant,
+    /// last timestamp of trying hole punch
+    /// it is used to send NatPingRequest every 3 seconds
+    pub last_punching_time: Instant,
+    /// factor variable for guessing NAT port
+    pub first_punching_index: u32,
+    /// another factor variable for guessing NAT port
+    pub last_punching_index: u32,
+    /// holds ping_id used in NatPingRequest
+    /// multi NatPingRequest has this same ping_id
+    /// because every NatPingRequest receives NatPingResponse.
+    pub ping_id: u64,
+}
+
+impl HolePunching {
+    /// new object of HolePunching
+    pub fn new() -> Self {
+        HolePunching {
+            is_punching_done: true,
+            num_punch_tries: 0,
+            last_recv_ping_time: Instant::now(),
+            last_send_ping_time: Instant::now(),
+            last_punching_time: Instant::now(),
+            first_punching_index: 0,
+            last_punching_index: 0,
+            ping_id: HolePunching::new_ping_id(),
```

Comment:
Because ping_id for NatPingRequest is parmanent per friend.
Send many NatPingRequest packet with same ping_id, one of them will be responded with the ping_id.

Reply:
But you don't send request when create `HolePunching` struct but set random ping_id. It seems nobody will know about it, and initial value can be zero.

Reply:
All ping_id of many NatPingRequest packets of a friend is same.
If ping_id should be set when DHT node send packet, it should be checked that this is first time of sending packet or not.

Reply:
Is there other good way?

Reply:
ok, got it.